src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1193,45 +1193,49 @@
                 :as report} (normalize-report orig-report)
                 report-hash (shash/report-identity-hash report)]
            (jdbc/with-db-transaction []
-             (let [certname-id (certname-id certname)
-                   row-map {:hash (sutils/munge-hash-for-storage report-hash)
-                            :transaction_uuid (sutils/munge-uuid-for-storage transaction_uuid)
-                            :catalog_uuid (sutils/munge-uuid-for-storage catalog_uuid)
-                            :code_id code_id
-                            :cached_catalog_status cached_catalog_status
-                            :metrics (sutils/munge-jsonb-for-storage metrics)
-                            :logs (sutils/munge-jsonb-for-storage logs)
-                            :resources (sutils/munge-jsonb-for-storage resources)
-                            :corrective_change corrective_change
-                            :noop noop
-                            :noop_pending noop_pending
-                            :puppet_version puppet_version
-                            :certname certname
-                            :report_format report_format
-                            :configuration_version configuration_version
-                            :producer_id (ensure-producer producer)
-                            :producer_timestamp producer_timestamp
-                            :start_time start_time
-                            :end_time end_time
-                            :receive_time (to-timestamp received-timestamp)
-                            :environment_id (ensure-environment environment)
-                            :status_id (ensure-status status)}
-                   [{report-id :id}] (->> row-map
-                                          maybe-environment
-                                          maybe-resources
-                                          maybe-corrective-change
-                                          (jdbc/insert! :reports))
-                   adjust-event-metadata #(-> %
-                                              (assoc :report_id report-id
-                                                     :certname_id certname-id)
-                                              maybe-corrective-change)]
-               (when-not (empty? resource_events)
-                 (->> resource_events
-                      (sp/transform [sp/ALL :containment_path] #(some-> % sutils/to-jdbc-varchar-array))
-                      (map adjust-event-metadata)
-                      (apply jdbc/insert! :resource_events)))
-               (when update-latest-report?
-                 (update-latest-report! certname)))))))
+             (let [shash (sutils/munge-hash-for-storage report-hash)]
+               (when-not (-> "select 1 from reports where hash = ? limit 1"
```

Comment:
I can see three ways of solving this problem

1. Check if the report exists already
2. Catch uniqueness constraint violations
3. Both

It seems like 1 falls short in the case where the same report is submitted twice and might be processed at the same time. Seems unlikely in the real world, possible in our tests. 2 would waste some time processing a report to only get a failure and ignore it (discard the report) but would catch the case I mentioned earlier. 3 would be more efficient in the common ways this error can occur and also protects us on concurrent submissions.